server/src/main/java/org/elasticsearch/plugins/UberModuleClassLoader.java

```diff
@@ -58,43 +59,66 @@ public class UberModuleClassLoader extends SecureClassLoader implements AutoClos
     private final ModuleLayer.Controller moduleController;
     private final Set<String> packageNames;
 
-    private static Map<String, Set<String>> getModuleToServiceMap(ModuleLayer moduleLayer) {
-        Set<String> unqualifiedExports = moduleLayer.modules()
+    private static Set<String> unqualifiedExportsForLayer(ModuleLayer layer) {
+        return layer.modules()
             .stream()
             .flatMap(module -> module.getDescriptor().exports().stream())
             .filter(Predicate.not(ModuleDescriptor.Exports::isQualified))
             .map(ModuleDescriptor.Exports::source)
             .collect(Collectors.toSet());
-        return moduleLayer.modules()
+    }
+
+    private static Set<ModuleDescriptor> readableModulesForLayer(ModuleLayer layer) {
+        return layer.modules()
             .stream()
             .map(Module::getDescriptor)
-            .filter(ModuleSupport::hasAtLeastOneUnqualifiedExport)
+            .filter(md -> ModuleSupport.hasAtLeastOneUnqualifiedExport(md) || md.isOpen() || md.isAutomatic())
+            .collect(Collectors.toSet());
+    }
+
+    private static Map<String, Set<String>> getModuleToServiceMap(List<ModuleLayer> moduleLayers) {
+        Set<String> allUnqualifiedExports = new HashSet<>();
+        Set<ModuleDescriptor> allReadableModules = new HashSet<>();
+        for (var layer : moduleLayers) {
+            allUnqualifiedExports.addAll(unqualifiedExportsForLayer(layer));
+            allReadableModules.addAll(readableModulesForLayer(layer));
+            for (var parentLayer : layer.parents()) {
+                // ModuleLayer.boot() contains all ES libs too, so we want it to be an explicit parent
+                if (parentLayer != ModuleLayer.boot()) {
+                    allUnqualifiedExports.addAll(unqualifiedExportsForLayer(parentLayer));
+                    allReadableModules.addAll(readableModulesForLayer(parentLayer));
+                    // TODO: recurse?
```

Comment:
We might want to make this fully recursive... maybe? 
(right now it does not seem we ever have extensions of extensions, but...)